test/cpp/nativert/test_mpmc_queue.cpp

```diff
@@ -0,0 +1,121 @@
+#include <atomic>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include <torch/nativert/detail/MPMCQueue.h>
+
+using torch::nativert::detail::MPMCQueue;
+
+TEST(MPMCQueueTest, EmptyQueue) {
+  MPMCQueue<int> queue(5);
+  int out = 0;
+  EXPECT_FALSE(queue.readIfNotEmpty(out));
+}
+
+TEST(MPMCQueueTest, SingleElement) {
+  MPMCQueue<int> queue(5);
+  EXPECT_TRUE(queue.writeIfNotFull(10));
+  int out = 0;
+  EXPECT_TRUE(queue.readIfNotEmpty(out));
+  EXPECT_EQ(out, 10);
+}
+
+TEST(MPMCQueueTest, MultipleElements) {
+  MPMCQueue<int> queue(5);
+  for (int i = 0; i < 5; ++i) {
+    EXPECT_TRUE(queue.writeIfNotFull(i));
+  }
+  for (int i = 0; i < 5; ++i) {
+    int out = 0;
+    EXPECT_TRUE(queue.readIfNotEmpty(out));
+    EXPECT_EQ(out, i);
+  }
+}
+
+TEST(MPMCQueueTest, FullQueue) {
+  MPMCQueue<int> queue(5);
+  for (int i = 0; i < 5; ++i) {
+    EXPECT_TRUE(queue.writeIfNotFull(i));
+  }
+  EXPECT_FALSE(queue.writeIfNotFull(10));
+}
+
+TEST(MPMCQueueTest, ConcurrentAccess) {
+  MPMCQueue<int> queue(10);
+  std::thread writer([&queue]() {
+    for (int i = 0; i < 5; ++i) {
+      queue.writeIfNotFull(i);
+    }
+  });
+  std::thread reader([&queue]() {
+    for (int i = 0; i < 5; ++i) {
+      int out = 0;
+      while (!queue.readIfNotEmpty(out)) {
+        // Wait until an element is available
+      }
```

Comment:
this is a bad pattern and we should not use it without great reason; you need to do something to yield the CPU in a loop like this

Reply:
yeah indeed, I can send a followup to address this.